pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/segment/writer/SegmentWriter.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.ingestion.segment.writer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import org.apache.pinot.spi.data.readers.GenericRow;
+
+
+/**
+ * An interface to collect records and create a Pinot segment.
+ * This interface helps abstract out details regarding segment generation from the caller.
+ */
+public interface SegmentWriter extends Closeable {
+
+  /**
+   * Initializes the {@link SegmentWriter} with provided tableConfig and Pinot schema.
+   */
+  void init(SegmentWriterConfig segmentWriterConfig)
+      throws Exception;
+
+  /**
+   * Collects a single {@link GenericRow} into a buffer.
+   * This row is not available in the segment until a <code>flush()</code> is invoked.
+   */
+  void collect(GenericRow row)
+      throws IOException;
+
+  /**
+   * Collects a batch of {@link GenericRow}s into a buffer.
+   * These rows are not available in the segment until a <code>flush()</code> is invoked.
+   */
+  void collect(GenericRow[] rowBatch)
+      throws IOException;
+
+  /**
+   * Creates one Pinot segment using the {@link GenericRow}s collected in the buffer,
+   * at the outputDirUri as specified in the batchConfigs.
```

Comment:
will the flush upload the segment?

Reply:
It will not. Upload will be done by calling SegmentUploader#upload.
We could introduce a wrapper class, that does segment generation and push, by calling SegmentWriter and SegmentUploader internally

Reply:
+1 to the wrapper class. Ideally we don't have to expose the segment files to the Spark/Flink connectors. 
Also, can you describe the flow of how the connector invokes this? 
Does it first call flush, and then call upload? I feel it's a bit unclean as `flush` does not return a handle, so how does the caller know what will be uploaded?

Previously I thought upload will also take care of the upload (ideally in an async way done by a thread), so the caller does not need to worry about the segment creation/upload details

Reply:
Yes, the caller will first call `SegmentWriter#flush()` and then `SegmentUploader#upload(segmentDir)`.
The segment will be generated in the `outputDirURI` that is provided in the tableConfig or in the batchConfigOverride provided by the caller. We can make `flush` return the segmentURI if you think that's better?

Yes previously with just one interface, flush would be doing both. But the aim with the new interface is to keep segment generation and segment push steps separate. There could be cases where user only wants to generate segment (e.g. preprocessing). Caller can use the wrapper class if they wish not to deal with 2 interfaces.


I'm not sure about async. What happens if the segment generation/push fails, or the thread gets killed, but flink has already started writing to the next segment? Who will be responsible for going back and rewriting the rows for that failed segment?

Reply:
I think returning `segmentURI ` from `flush` will be good, so as you said the caller can decide on uploading some specific segments, instead of all under the folder.

The async is just the wrapper implementation, it shall package some default fault-tolerance mechanisms to handle generation/push fails with retry etc. Async just means that the caller can decide to run it in a separate thread and not to block `segmentWriter.collect`